#159993338 Login via Google and Facebook

hoslack opened the pull request:
#### What does this PR do?
- Enables user to login with Facebook and Google

#### Description of Task to be completed?
- Create facebook and google login buttons
- Get the tokens from the auth providers above
- Send the auth token from the providers to the backend API for verification using a redux action
- Update the reducer with user details after getting verification from the backend 
- Update the localstorage with current user details
- Redirect the user to the home page

#### How should this be manually tested?
- Git pull the source code from ft-Login-via-Google-Facebook-Twitter-159993338 branch
- Copy the variables from sample.env and put the correct parameters in .env file
- Run `npm i` to install the dependencies used in the project
- Run `npm start` to start the project
- Go to the browser and try loging in with Facebook and Google
#### What are the relevant pivotal tracker stories?
[#159993338](https://www.pivotaltracker.com/n/projects/2184741/stories/159993338)
#### Screenshots (if appropriate)

- Login Page
<img width="1220" alt="screen shot 2018-09-05 at 00 57 06" src="https://user-images.githubusercontent.com/15017978/45060013-d89c9400-b0a6-11e8-9b07-5b712d934772.png">

- During Login
<img width="1217" alt="screen shot 2018-09-05 at 00 57 26" src="https://user-images.githubusercontent.com/15017978/45060015-da665780-b0a6-11e8-974e-539880772c4d.png">

- After Login
<img width="1226" alt="screen shot 2018-09-05 at 00 57 46" src="https://user-images.githubusercontent.com/15017978/45060018-dc301b00-b0a6-11e8-9658-d8bf18a90f55.png">

hoslack committed:
[ch 159993338] Refactor code to use App as the main router

hoslack committed:
[ft 159993338] Add frontend function for google login

hoslack committed:
[fix 159993338] Fix google ID import error

hoslack committed:
[fix 159993338] fix import issues in config file

hoslack committed:
[fix 159993338] Use .env file for secret variables

hoslack committed:
[ft 159993338] Add google button that returns user access token to the front end

hoslack committed:
[ft 159993338] Add google auth with redux
Store user credentials in localstorage for persistence

hoslack committed:
[fix 159993338] Add general function for social login to make code DRY

davenjeru committed:
[Feature #159993338] add .env-sample

Signed-off-by: davenjeru <dave.njeru@andela.com>

davenjeru committed:
[Feature #159993338] get social url from enviroment variable

Signed-off-by: davenjeru <dave.njeru@andela.com>

davenjeru committed:
[Feature #159993338] add login with facebook

Signed-off-by: davenjeru <dave.njeru@andela.com>

davenjeru committed:
[Feature #159993338] fix error in props destructuring
 - remove destructuring of props since the 'SOCIALAUTH' was not found
   when destructuring was used.

Signed-off-by: davenjeru <dave.njeru@andela.com>

davenjeru committed:
Merge branch 'ft-login-page-159993337' of https://github.com/andela/ah-scorpion-frontend into demo

davenjeru committed:
Merge branch 'ft-Login-via-Google-Facebook-Twitter-159993338' of https://github.com/andela/ah-scorpion-frontend into demo

davenjeru committed:
ready for demo

davenjeru committed:
[Feature #159993338] rebase this branch on top 'ft-login-page-159993337'

davenjeru committed:
[Feature #159993338] remove 'autoLoad' prop from FacebookLogin

Signed-off-by: davenjeru <dave.njeru@andela.com>

davenjeru committed:
[Feature #159993338] redirect to home on successful social login

Signed-off-by: davenjeru <dave.njeru@andela.com>

davenjeru committed:
[Feature #159993338] use a 'Promise' to redirect to home on successful social login

Signed-off-by: davenjeru <dave.njeru@andela.com>

davenjeru committed:
[Chore #159993330] add loginReducer tests

Signed-off-by: davenjeru <dave.njeru@andela.com>

davenjeru committed:
[Chore #159993330] add 'Footer' in Login.js

Signed-off-by: davenjeru <dave.njeru@andela.com>

hoslack committed:
Add tests to social auth functionality
# Conflicts:
#	package-lock.json
#	src/containers/Login.js

hoslack committed:
[fix 159993338] remove double quotes from code

hoslack committed:
[fix 159993338] fix import/export issue in eslint

hoslack committed:
[fix 159993338] edit eslint config file to allow import/export of modules

hoslack committed:
[fix 159993338] Add eslint config file to hound config

RuthNjeri commented on src/actions/types.js:
```diff
@@ -1 +1,4 @@
-export const TEST_REDUX = 'TEST_REDUX';
+export const BEGIN_LOGIN = 'BEGIN_LOGIN';
```
Add `.eslintrc.json`  in the Hound yml file for it to detect the eslint configurations.

hoslack replied:
I have just done that @RuthNjeri . Let me see the new results

davenjeru committed:
[Chore #159993330] modify .eslintrc.json
 - add 'import' under plugins
 - add "browser: true" under env
 - add "allowImportExportEverywhere": true under parserOptions
 - install redux-devtools-extension in dependencies instead of
   devDependencies

Signed-off-by: davenjeru <dave.njeru@andela.com>

davenjeru commented on src/reducers/loginReducer.js:
```diff
@@ -0,0 +1,36 @@
+import * as types from '../actions/types';
```
Added "babel-eslint" as parser and editted the "parserOptions" to allow "sorceType:  module"

davenjeru commented on src/index.js:
```diff
@@ -1,16 +1,22 @@
     <App />
   </Provider>,
-  document.getElementById("root")
+  document.getElementById('root'),
```
Added "browser: true" in "env:{}" to allow the use of the 'document' keyword.

davenjeru commented on src/components/LoginForm.js:
```diff
@@ -0,0 +1,118 @@
+      this.setState({ loading: true });
+      this.props
+        .submit(this.state.data)
+        .catch(err => this.setState({ errors: err.response.data.errors, loading: false }));
```
@blac-siren should take care of LoginForm.js

davenjeru committed:
[Chore #159993330] fix hound violations
 - export social login as default from 'actions/index.js'
 - used array destructuring to get the error message
 - import default socialLogin from 'actions/'
 - used props destructuring in the constructor
 - remove console statement from onFailure()

Signed-off-by: davenjeru <dave.njeru@andela.com>

davenjeru committed:
[Chore #159993330] fix code climate issues
 - remove a pair of div tags, now render has 25 or less lines of code
 - escape "\'" character in HTML statement
 - add type="submit" in login button

Signed-off-by: davenjeru <dave.njeru@andela.com>

davenjeru committed:
[Chore #159993330] configure eslint to ignore .js import extensions

Signed-off-by: davenjeru <dave.njeru@andela.com>

davenjeru committed:
[Chore #159993330] configure eslint to allow consistent line breaks on
parentheses

Signed-off-by: davenjeru <dave.njeru@andela.com>

davenjeru committed:
[Chore #159993330] correct 'import/extensions' configuration in eslint

Signed-off-by: davenjeru <dave.njeru@andela.com>

davenjeru committed:
[Chore #159993330] allow import extensions of .css files

Signed-off-by: davenjeru <dave.njeru@andela.com>

davenjeru committed:
[Chore #159993330] turn 'function-paren-newline' off

Signed-off-by: davenjeru <dave.njeru@andela.com>

davenjeru committed:
[Chore #159993330] add '/signup' to the Link component in LoginForm.js

Signed-off-by: davenjeru <dave.njeru@andela.com>

davenjeru committed:
Merge branch 'ch-Authentication-160311539' of https://github.com/andela/ah-scorpion-frontend into ft-Login-via-Google-Facebook-Twitter-159993338

davenjeru committed:
[Chore #159993330] remove map state to props

Signed-off-by: davenjeru <dave.njeru@andela.com>
